gapis/resolve/dependencygraph/footprint.go

```diff
@@ -0,0 +1,224 @@
+// executing the command to whom it belongs. Behaviour also contains a refernce
+// to the back-propagation machine which should be used to process the
+// Behaviour to determine its liveness for dead code elimination.
+type Behaviour struct {
+	Reads    []DefUseAtom
+	Writes   []DefUseAtom
+	BelongTo api.SubCmdIdx
+	Alive    bool
+	Aborted  bool
+	Machine  BackPropagationMachine
+}
+
+// NewBehaviour creates a new Behaviour which belongs to the command indexed by
+// the given SubCmdIdx and shall be process by the given back-propagation
+// machine. Returns a pointer to the created Behaviour.
+func NewBehaviour(fullCommandIndex api.SubCmdIdx, m BackPropagationMachine) *Behaviour {
+	return &Behaviour{
+		BelongTo: fullCommandIndex,
+		Machine:  m,
+	}
+}
+
+// Read records a read operation of the given DefUseAtom to the Behaviour
+func (b *Behaviour) Read(c DefUseAtom) {
+	b.Reads = append(b.Reads, c)
+}
+
+// Write records a write operation of the given DefUseAtom to the Behaviour
+func (b *Behaviour) Write(c DefUseAtom) {
+	b.Writes = append(b.Writes, c)
+}
+
+// Modify records a read and a write operation of the given DefUseAtom to the
+// Behaviour
+func (b *Behaviour) Modify(c DefUseAtom) {
+	b.Read(c)
+	b.Write(c)
+}
+
+// DefUseAtom represents a piece of state which can be altered as commands
+// being executed.
+type DefUseAtom interface {
+	DefUseAtom()
+}
+
+// BackPropagationMachine determines the liveness of Behaviours along the
+// back propagation of the operations recorded in Footprint's Behaviours.
+type BackPropagationMachine interface {
+	// IsAlive check whether a given behaviour, which is specified by its index
+	// and the footprint to which it belongs, should be kept alive. Returns true
+	// if the behaviour should be kept alive, otherwise returns false.
+	IsAlive(behaviourIndex uint64, ft *Footprint) bool
+	// RecordBehaviourEffects records the read/write operations of the given
+	// behaviour specified by its index in the given footprint, and returns the
+	// indices of the alive behaviours that should be kept alive due to recording
+	// of the given behaviour.
+	RecordBehaviourEffects(behaviourIndex uint64, ft *Footprint) []uint64
+	// Clear clears the internal state of the BackPropagationMachine.
+	Clear()
+}
+
+// dummyMachine does nothing but marks all the incoming Behaviours as alive.
+type dummyMachine struct{}
+
+func (m *dummyMachine) IsAlive(uint64, *Footprint) bool { return true }
+func (m *dummyMachine) RecordBehaviourEffects(behaviourIndex uint64,
+	ft *Footprint) []uint64 {
+	return []uint64{behaviourIndex}
+}
+
+func (m *dummyMachine) Clear() {}
+
+// GetBehaviourIndex returns the index of the last Behaviour in the Footprint
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or a valid Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
```

Comment:
Unless there's a horrible variable collision, please use single letter method receivers.
Please try to keep function declarations on a single line unless they're crazy long - I find the newline hard to separate from the body.
Also try to avoid `Get` prefixes:

`func (f *Footprint) BehaviorIndex(ctx context.Context, i api.SubCmdIdx) uint64 {`